File: src/main/java/club/xiaojiawei/starter/GameStarter.java
package club.xiaojiawei.starter;

import club.xiaojiawei.interfaces.closer.StarterTaskCloser;
import club.xiaojiawei.custom.LogRunnable;
import club.xiaojiawei.data.ScriptStaticData;
import club.xiaojiawei.utils.GameUtil;
import club.xiaojiawei.utils.MouseUtil;
import club.xiaojiawei.utils.RandomUtil;
import club.xiaojiawei.utils.SystemUtil;
import com.sun.jna.platform.win32.User32;
import com.sun.jna.platform.win32.WinDef;
import jakarta.annotation.Resource;
import javafx.beans.property.BooleanProperty;
import lombok.extern.slf4j.Slf4j;
import org.springframework.context.annotation.Lazy;
import org.springframework.stereotype.Component;

import java.awt.*;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.ScheduledThreadPoolExecutor;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicReference;

/**
 * 启动游戏
 * @author 肖嘉威
 * @date 2023/7/5 14:38
 */
@Slf4j
@Component
public class GameStarter extends AbstractStarter implements StarterTaskCloser {

    @Resource
    private AtomicReference<BooleanProperty> isPause;
    @Resource
    private ScheduledThreadPoolExecutor launchProgramThreadPool;
    @Resource
    private MouseUtil mouseUtil;
    @Resource
    private ScheduledThreadPoolExecutor extraThreadPool;
    @Lazy
    @Resource
    private AbstractStarter starter;
    @Resource
    private GameUtil gameUtil;

    private ScheduledFuture<?> scheduledFuture;

    private WinDef.HWND gameHWND;

    @Override
    public void exec() {
        log.info("开始检查" + ScriptStaticData.GAME_CN_NAME);
        if ((gameHWND = SystemUtil.findGameHWND()) != null){
            cancelAndStartNext();
            return;
        }
        final AtomicInteger launchCount = new AtomicInteger();
        scheduledFuture = launchProgramThreadPool.scheduleAtFixedRate(new LogRunnable(() -> {
            if (isPause.get().get()) {
                cancelGameTimer();
            } else {
                if (launchCount.incrementAndGet() > 3){
                    user32LaunchGame();
                }else if (launchCount.incrementAndGet() > 4){
                    log.warn("打开炉石失败次数过多，重新执行启动器链");
                    cancelGameTimer();
                    extraThreadPool.schedule(() -> {
                        SystemUtil.killLoginPlatform();
                        SystemUtil.killPlatform();
                        launchCount.set(0);
                        starter.start();
                    }, 1, TimeUnit.SECONDS);
                    return;
                }
                if (SystemUtil.isAliveOfGame()) {
//                    游戏刚启动时找不到HWND
                    if ((gameHWND = SystemUtil.findGameHWND()) == null){
                        log.info("炉石传说已在运行，但未找到对应窗口句柄");
                        return;
                    }
                    cancelAndStartNext();
                }else {
                    GameUtil.cmdLaunchGame();
                }
            }
        }), 5, 20, TimeUnit.SECONDS);
    }

    private void user32LaunchGame(){
        log.info("正在通过user32打开" + ScriptStaticData.GAME_CN_NAME);
        WinDef.HWND platformhwnd = SystemUtil.findPlatformHWND();
        MouseUtil.leftButtonClick(new Point(145, 120), platformhwnd);
    }

    private void cancelGameTimer(){
        if (scheduledFuture != null && !scheduledFuture.isDone()){
            scheduledFuture.cancel(true);
        }
    }

    private void cancelAndStartNext(){
        log.info(ScriptStaticData.GAME_CN_NAME + "正在运行");
        cancelGameTimer();
        GameUtil.hidePlatformWindow();
        extraThreadPool.schedule(() -> {
            ScriptStaticData.setGameHWND(gameHWND);
            SystemUtil.updateGameRect();
            startNextStarter();
        },1, TimeUnit.SECONDS);
    }

    @Override
    public void closeStarterTask() {
        cancelGameTimer();
    }

    @Override
    public int getOrder() {
        return 60;
    }
}

File: Hearthstone-Script/src/main/java/club/xiaojiawei/hsscript/starter/GameStarter.kt
package club.xiaojiawei.hsscript.starter

import club.xiaojiawei.bean.LRunnable
import club.xiaojiawei.config.EXTRA_THREAD_POOL
import club.xiaojiawei.config.LAUNCH_PROGRAM_THREAD_POOL
import club.xiaojiawei.config.log
import club.xiaojiawei.hsscript.config.StarterConfig
import club.xiaojiawei.hsscript.data.GAME_CN_NAME
import club.xiaojiawei.hsscript.data.GAME_HWND
import club.xiaojiawei.hsscript.dll.SystemDll
import club.xiaojiawei.hsscript.enums.MouseControlModeEnum
import club.xiaojiawei.hsscript.status.PauseStatus
import club.xiaojiawei.hsscript.utils.GameUtil
import club.xiaojiawei.hsscript.utils.MouseUtil
import club.xiaojiawei.hsscript.utils.SystemUtil
import com.sun.jna.platform.win32.WinDef
import com.sun.jna.platform.win32.WinDef.HWND
import java.awt.Point
import java.io.File
import java.util.concurrent.TimeUnit

/**
 * 启动游戏
 * @author 肖嘉威
 * @date 2023/7/5 14:38
 */
class GameStarter : AbstractStarter() {

    private var latestLogDir: File? = null

    public override fun execStart() {
        log.info { "开始检查$GAME_CN_NAME" }
        GameUtil.findGameHWND()?.let {
            next(it)
            return
        }
        latestLogDir = GameUtil.getLatestLogDir()
        var startTime = System.currentTimeMillis()
        var firstLogLaunch = true
        var firstLogSecondaryLaunch = true
        addTask(
            LAUNCH_PROGRAM_THREAD_POOL.scheduleAtFixedRate(LRunnable {
                val diffTime = System.currentTimeMillis() - startTime
                if (diffTime > 20_000) {
                    log.warn { "启动${GAME_CN_NAME}失败次数过多，重新执行启动器链" }
                    stopTask()
                    startTime = System.currentTimeMillis()
                    EXTRA_THREAD_POOL.schedule({
                        GameUtil.killGame(true)
                        GameUtil.killLoginPlatform()
                        GameUtil.killPlatform()
                        StarterConfig.starter.start()
                    }, 1, TimeUnit.SECONDS)
                    return@LRunnable
                } else if (diffTime > 5_000) {
                    if (firstLogSecondaryLaunch){
                        firstLogSecondaryLaunch = false
                        log.info { "更改${GAME_CN_NAME}启动方式" }
                    }
                    GameUtil.launchPlatformAndGame()
                }
                if (GameUtil.isAliveOfGame()) {
//                    游戏刚启动时可能找不到窗口句柄
                    GameUtil.findGameHWND()?.let {
                        next(it)
                    } ?: let {
                        if (diffTime > 5_000) {
                            log.info { "${GAME_CN_NAME}已在运行，但未找到对应窗口句柄" }
                            return@LRunnable
                        }
                    }
                } else {
                    if (firstLogLaunch){
                        firstLogLaunch = false
                        log.info { "正在启动$GAME_CN_NAME" }
                    }
                    launchGameBySendMessage()
                }
            }, 100, 500, TimeUnit.MILLISECONDS)
        )
    }

    private fun launchGameBySendMessage() {
        val platformHWND = GameUtil.findPlatformHWND()
        val rect = WinDef.RECT()
        SystemUtil.updateRECT(platformHWND, rect)
        MouseUtil.leftButtonClick(
            Point(145, rect.bottom - rect.top - 150),
            platformHWND,
            MouseControlModeEnum.MESSAGE.code
        )
    }

    private fun next(gameHWND: HWND) {
        log.info { GAME_CN_NAME + "正在运行" }
        if (latestLogDir != null) {
            log.info { "等待${GAME_CN_NAME}创建最新日志文件夹" }
            while (!PauseStatus.isPause) {
                val currentLatestLogDir = GameUtil.getLatestLogDir()
                if (currentLatestLogDir != null) {
                    if (currentLatestLogDir > latestLogDir){
                        log.info { "${GAME_CN_NAME}已创建最新日志文件夹：${currentLatestLogDir.absolutePath}" }
                        break
                    }
                }
                Thread.sleep(200)
            }
        }
        updateGameMsg(gameHWND)
        startNextStarter()
    }

    private fun updateGameMsg(gameHWND: HWND) {
        GAME_HWND = gameHWND
        GameUtil.updateGameRect()
        Thread.ofVirtual().name("Update GameRect VThread").start {
            Thread.sleep(3000)
            GameUtil.updateGameRect()
            SystemDll.INSTANCE.changeWindow(GAME_HWND, true)
            SystemDll.INSTANCE.changeInput(GAME_HWND, true)
        }
    }

}
